common/src/main/java/com/microsoft/identity/common/internal/ui/webview/challengehandlers/ClientCertAuthChallengeHandler.java

```diff
@@ -22,64 +22,434 @@
+
+    //Helper method that returns a list of YubiKitCertDetails extracted from certificates located on the YubiKey.
+    //This method should only be called within a callback upon creating a successful YubiKey device connection.
+    private List<YubiKitCertDetails> getCertDetailsFromKey(@NonNull final PivSession piv) {
+        //Create ArrayList that contains cert details only pertinent to the cert picker.
+        List<YubiKitCertDetails> certList = new ArrayList<>();
+        //We need to check every slot. If a slot is empty, an ApduException is called, in which case, we'll ignore it.
+        //AUTHENTICATION (9A)
+        getAndPutCertDetailsInList(AUTHENTICATION, piv, certList);
+        //SIGNATURE (9C)
+        getAndPutCertDetailsInList(SIGNATURE, piv, certList);
+        //KEY_MANAGEMENT (9D)
+        getAndPutCertDetailsInList(KEY_MANAGEMENT, piv, certList);
+        //CARD_AUTH (9E)
+        getAndPutCertDetailsInList(CARD_AUTH, piv, certList);
+
+        return certList;
+    }
+
+    //Helper method created to handle reading certificates off YubiKey.
+    //This method should only be called within a callback upon creating a successful YubiKey device connection.
+    private void getAndPutCertDetailsInList(@NonNull final Slot slot, @NonNull final PivSession piv, @NonNull final List<YubiKitCertDetails> certList) {
+        final String methodTag = TAG + ":getAndPutCertDetailsInList";
+        try {
+            final X509Certificate cert =  piv.getCertificate(slot);
+            //If there are no exceptions, add details of this cert to our certList.
+            certList.add(new YubiKitCertDetails(cert.getIssuerDN().getName(), cert.getSubjectDN().getName(), slot));
+        } catch (IOException e) {
+            Logger.errorPII(methodTag,"IOException", e);
+        } catch (BadResponseException e) {
+            Logger.errorPII(methodTag,"BadResponseException", e);
+        } catch (ApduException e) {
+            //If sw is 0x6a82 (27266), This is a FILE_NOT_FOUND error, which we should ignore since this means the slot is merely empty.
+            if (e.getSw() != 0x6a82) {
+                Logger.errorPII(methodTag,"ApduException", e);
+            }
+        }
+    }
+    //Build and show Smartcard cert picker, which also handles the rest of the smartcard CBA flow.
+    private void showSmartcardCertPickerDialog(@NonNull final List<YubiKitCertDetails> certList, @NonNull final ClientCertRequest request) {
+        final SmartcardCertPickerDialog certPickerDialog = new SmartcardCertPickerDialog(certList, mActivity);
+        certPickerDialog.setPositiveButtonListener(getSmartcardCertPickerDialogPositiveButtonListener(request));
+        certPickerDialog.setNegativeButtonListener(new SmartcardCertPickerDialog.NegativeButtonListener() {
+            @RequiresApi(api = Build.VERSION_CODES.LOLLIPOP)
+            @Override
+            public void onClick() {
+                request.cancel();
+                mCurrentDialog = null;
```

Comment:
We do need to set `mCurrentDialog` to null because the negative button listener of the dialog will automatically dismiss the dialog. 
If the yubikey is unplugged after a user hits "cancel", the `setOnClosed` callback will check if mCurrentDialog is equal to null. If it isn't null, it will attempt to dismiss the dialog.  I believe attempting to dismiss a dialog twice will throw some sort of error.